pkg/utils/reconciler.go

```diff
@@ -700,3 +700,41 @@ func (r *ReconcilerBase) GetRouteTLSValues(ba common.BaseApplication) (key strin
 	}
 	return key, cert, ca, destCa, nil
 }
+
+// GetSelectorLabelsFromApplications finds application CRs with the specified name in the BaseApplication's namespace and returns labels in `selector.matchLabels`.
+// If it fails to find in the current namespace, it looks up in the whole cluster and aggregates all labels in `selector.matchLabels`.
+func (r *ReconcilerBase) GetSelectorLabelsFromApplications(ba common.BaseApplication) (map[string]string, error) {
+	mObj := ba.(metav1.Object)
+	allSelectorLabels := map[string]string{}
+	app := &applicationsv1beta1.Application{}
+	key := types.NamespacedName{Name: ba.GetApplicationName(), Namespace: mObj.GetNamespace()}
+	var err error
+	if err = r.GetClient().Get(context.Background(), key, app); err == nil {
+		if app.Spec.Selector != nil {
+			for name, value := range app.Spec.Selector.MatchLabels {
+				allSelectorLabels[name] = value
+			}
+		}
+	} else if err != nil && kerrors.IsNotFound(err) {
+		apps := &applicationsv1beta1.ApplicationList{}
+		if err = r.GetClient().List(context.Background(), apps, client.InNamespace("")); err == nil {
```

Comment:
If intention for ApplicationName to be used as a name of app to join, we should be checking that here as well. It could be that there are 2+ apps that have `kappnav.component.namespaces`  annotation with same NS list. And I think this code will copy  the labels from all of them

Reply:
Good catch! 🥇